src/passes/InstrumentBranchHints.cpp

```diff
@@ -0,0 +1,334 @@
+/*
+ * Copyright 2025 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Instruments branch hints and their targets, adding logging that allows us to
+// see if the hints were valid or not. We turn
+//
+//  @metadata.branch.hint B
+//  if (condition) {
+//    X
+//  } else {
+//    Y
+//  }
+//
+// into
+//
+//  @metadata.branch.hint B
+//  ;; log the ID of the condition (123), the prediction (B), and the actual
+//  ;; runtime result (temp == condition).
+//  if (temp = condition; log(123, B, temp); temp) {
+//    X
+//  } else {
+//    Y
+//  }
+//
+// Concretely, we emit calls to this logging function:
+//
+//  (import "fuzzing-support" "log-branch"
+//    (func $log-branch (param i32 i32 i32)) ;; ID, prediction, actual
+//  )
+//
+// This can be used to verify that branch hints are accurate, by implementing
+// the import like this for example:
+//
+//  imports['fuzzing-support']['log-branch'] = (id, prediction, actual) => {
+//    // We only care about truthiness of the expected and actual values.
+//    expected = +!!expected;
+//    actual = +!!actual;
+//    // Throw if the hint said this branch would be taken, but it was not, or
+//    // vice versa.
+//    if (expected != actual) throw `Bad branch hint! (${id})`;
+//  };
+//
+// Another use case for this pass is to fuzz branch hint updates: given a fuzz
+// case, we can instrument it and view the loggings, then optimize the original,
+// instrument that, and view those loggings. Imagine, for example, that we flip
+// the condition but forget to flip the hint:
+//
+//  @metadata.branch.hint B
+//  if (!(temp = condition; log(123, B, temp); temp)) { ;; added a !
```

Comment:
What's to stop the `!` from going on the final `temp` in the block? That seems like it would mess up the second instrumentation described below.

Reply:
I see, so there's no guarantee that we are able to emit the second instrumentation if the module has been modified after the first instrumentation? It might be useful to call that out in the top-level comment.